### PR TITLE
feat(24.04): add openssh-{client,server,sftp-server} and its dependencies

### PR DESCRIPTION
**Pull request by Meulengracht**
First PR of a longer chain to add all dependencies needed to create a rootfs for core bases, as a part of switching to Chisel for building the rootfs for UC26.

This has been tested as of 22/02 and works.

### CODE REVIEW

**Comment by Meulengracht** [comment]
@rebornplusplus @cjdcordeiro Updated PR

**Comment by rebornplusplus** [comment]
Hi Philip, will you kindly check the linting issues in the CI and the package dependencies in the above comment? I will review this shortly, before our sync on Friday.

**Comment by Meulengracht** [comment]
> Hi Philip, will you kindly check the linting issues in the CI and the package dependencies in the above comment? I will review this shortly, before our sync on Friday.

The linting issue is a bit annoying because I'm just writing a file verbatim from the package postinst script, I don't want to change it as the file itself is not wide (and matches what the package writes), but due to the indentation it gets a bit wider than what the linter accepts.

Could we ignore it?

**Comment by rebornplusplus** [comment]
> The linting issue is a bit annoying because I'm just writing a file verbatim from the package postinst script, I don't want to change it as the file itself is not wide (and matches what the package writes), but due to the indentation it gets a bit wider than what the linter accepts.
> 
> Could we ignore it?

Hmm... we should ignore it, yeah. I will make sure to change the line-length. But since the CI is skipping another linting check, could you please run the next check locally and see if something's off?

<details>
<summary>Linter script</summary>

```sh
#!/bin/bash

set -e

export LC_COLLATE=C
EXIT_CODE=0

err() {
  echo "error:   " "$@" >&2
  EXIT_CODE=1
}
warn() {
  echo "warning: " "$@" >&2
}

for filename in $(find slices/ | grep "\.yaml$" | sort); do
  yq '.slices | keys | .[]' "$filename" | sort -C || \
    warn "$filename: slice names are not sorted"
  for slice in $(yq '.slices | keys | .[]' "$filename"); do
    key="$slice" yq \
      '.slices | with_entries(select(.key == env(key))) | .[].essential[]' \
      "$filename" | sort -C || \
      err "$filename: $slice: \"essential\" entries are not sorted"
    key="$slice" yq \
      '.slices | with_entries(select(.key == env(key))) | .[].contents | select(.) | keys | .[]' \
      "$filename" | sort -C || \
      err "$filename: $slice: \"contents\" entries are not sorted"
  done
done

exit $EXIT_CODE
```

</details>

Place the script within your chisel-releases directory and run it like following:

```
./linter-script
```

File: slices/sensible-utils.yaml
Comment (rebornplusplus) [inline slices/sensible-utils.yaml:18]:
```suggestion
  mime:
```

File: slices/ucf.yaml
Comment (rebornplusplus) [inline slices/ucf.yaml:13]: We do not need the mime file from `sensible-utils`.

```suggestion
      - sensible-utils_bins
```

File: slices/sensible-utils.yaml
Comment (rebornplusplus) [inline slices/sensible-utils.yaml:7]: It seems to me that we are using the `base` slices (in some other places too) to contain pretty much the standard almost-everything bits from a package. The problem here would be if we start listing the `base` slices as dependencies instead of something small which would meet the necessities, for example, `bins`. In that case our rootfs will contain stuff we do not really need.
Let me know what you think about this.

File: slices/ucf.yaml
Comment (rebornplusplus) [inline slices/ucf.yaml:22]: The following lines are from the `postinst` script. Do we need to create those as well?

```
    touch /var/lib/ucf/hashfile
    touch /var/lib/ucf/registry
```

**Review by rebornplusplus — COMMENTED** [review]
I went through the slices and they all are looking quite good actually. Great chiselling there, thanks! I only left a bunch of comments and suggestions that might improve the slices a bit.

File: slices/init-system-helpers.yaml
Comment (rebornplusplus) [reply slices/init-system-helpers.yaml]: This is outdated.

File: slices/openssh-server.yaml
Comment (rebornplusplus) [inline slices/openssh-server.yaml:29]: Do we need this one? I only noticed now that this is not listed as a dependency in the package control file.

**Comment by rebornplusplus** [comment]
Hiya @Meulengracht, about the CI, I have duplicated this PR in https://github.com/rebornplusplus/chisel-releases/pull/11. That repo has modified CI where line length issues are warnings, not errors. You do not need to do anything there, pushing your changes in this branch should update both PRs. Just make sure to check if there are any errors in that CI run.

File: slices/libpam-modules.yaml
Comment (Meulengracht) [reply slices/libpam-modules.yaml:74]: Files under `/usr/share/pam-configs` are shipped by various packages - It's loaded by pam-auth-update which actually is not used at all since pam-auth-update relies on debconf environment, which is not available. Instead in core we ship static pam files so we can control the contents.

I'll move these files to their own slice and not reference to them

File: slices/libselinux1.yaml
Comment (Meulengracht) [reply slices/libselinux1.yaml:11]: Nice idea

File: slices/openssh-client.yaml
Comment (Meulengracht) [reply slices/openssh-client.yaml:27]: They are not referenced by scripts, but rather used by the binaries

File: slices/ucf.yaml
Comment (Meulengracht) [reply slices/ucf.yaml:15]: Thank you! Is the syntax for empty files correct in this file?

File: slices/libpam-modules.yaml
Comment (Meulengracht) [reply slices/libpam-modules.yaml]: Corrected a lot of these dependencies, hope they look okay now

File: slices/libpam-runtime.yaml
Comment (Meulengracht) [reply slices/libpam-runtime.yaml]: I reorganized this file, please take a look again

File: slices/sensible-utils.yaml
Comment (Meulengracht) [reply slices/sensible-utils.yaml]: good suggestion!

File: slices/sensible-utils.yaml
Comment (Meulengracht) [reply slices/sensible-utils.yaml]: Updated this based on our discussion

File: slices/openssh-server.yaml
Comment (Meulengracht) [reply slices/openssh-server.yaml]: You are right, seems this has changed

File: slices/ucf.yaml
Comment (Meulengracht) [reply slices/ucf.yaml]: nice catch, updated this

File: slices/openssh-server.yaml
Comment (Meulengracht) [reply slices/openssh-server.yaml]: updated!

File: slices/openssh-client.yaml
Comment (rebornplusplus) [reply slices/openssh-client.yaml:27]: Oh alright. I thought it was the other way around. Thanks for letting me know!

File: slices/ucf.yaml
Comment (rebornplusplus) [reply slices/ucf.yaml:17]: It's good, thanks!

File: slices/libpam-runtime.yaml
Comment (rebornplusplus) [inline slices/libpam-runtime.yaml:19]: I am not sure if this is a better name or not. But `folders` sound a bit weird. Let me know if you have other suggestions!
```suggestion
  var:
```

**Review by rebornplusplus — CHANGES_REQUESTED** [review]
Thanks for addressing the previous comments! Looks much better now. 

I just have one minor nitpick about a slice name and please note that linting (excluding line-length) is failing.

**Review by rebornplusplus — APPROVED** [review]
Looks good to me. Thank you very much for all these slices!

This one will need to wait for another maintainer-approval. In the meantime, please feel free to open other PRs.

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
(already discussed - need a rebase and conflict resolution)

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
@Meulengracht congrats for the successful tests :) thanks for the changes.

Could you please also add the copyright slices to each SDF? We are now making these explicit for every package, to allow for a better traceability story.

**Comment by Meulengracht** [comment]
@cjdcordeiro @rebornplusplus is there more I can do?

File: slices/libcbor0.10.yaml
Comment (cjdcordeiro) [inline slices/libcbor0.10.yaml:11]: well done. you didn't fall into the temptation of doing `libcbor.so.0*` :stuck_out_tongue:

File: slices/liblzma5.yaml
Comment (cjdcordeiro) [inline slices/liblzma5.yaml:11]: good catch

File: slices/openssh-server.yaml
Comment (cjdcordeiro) [inline slices/openssh-server.yaml:20]: why is this needed? it's not in the deb's dependencies

File: slices/openssh-server.yaml
Comment (cjdcordeiro) [inline slices/openssh-server.yaml:55]: This is an interesting approach. 3 of these dependencies are not listed for the `bins` slice above. Furthermore, `libpam` is only referenced here. So my question is: if I just install `openssh-server_bins`, would it work?

File: slices/openssh-sftp-server.yaml
Comment (cjdcordeiro) [inline slices/openssh-sftp-server.yaml:14]: Same question here as above - if I install the `bins` slice alone, I won't get `openssh-client_standard`, so would those 2 bins be functional?

File: slices/procps.yaml
Comment (cjdcordeiro) [inline slices/procps.yaml:7]: see my comment in the review, but I think this `standard` might actually be a `core`?

File: slices/sensible-utils.yaml
Comment (cjdcordeiro) [inline slices/sensible-utils.yaml:14]: what about `/usr/share/sensible-utils/bin/realpath`? And, are these two really needed here or are they auxiliary binaries?

File: slices/sensible-utils.yaml
Comment (cjdcordeiro) [inline slices/sensible-utils.yaml:24]: yet another case where this time, it feels like `standard` is actually delivering "all".

I get the usefulness of the modularity, but given that `standard` and `core` are quite vague words, I'd refrain from using them unless they are actually delivering some standard functionality (like the `standard` slice in the `python3` slices`)

In this case (and others above) I'd suggest simply adding the `bins` and `mime` slices as dependencies (instead of the `standard` slice)

File: slices/sysvinit-utils.yaml
Comment (cjdcordeiro) [inline slices/sysvinit-utils.yaml:27]: exactly the same comment as the previous comment in `sensible-utils`

File: slices/sysvinit-utils.yaml
Comment (cjdcordeiro) [inline slices/sysvinit-utils.yaml:23]: can u move it down please?

File: slices/ucf.yaml
Comment (cjdcordeiro) [inline slices/ucf.yaml:28]: same as above

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Thanks a lot for this work @Meulengracht ! 

I think most comments are simply nits and design related. The actual contents look ok to me so we should be close to merge.

I feel like there's a bit of rethinking needed around the use of `core` and `standard`. Sometimes it feels like the `standard` is delivering just the "core" functionality, other times it feels like it's delivering everything in the SDF...

File: slices/libpam-runtime.yaml
Comment (Meulengracht) [reply slices/libpam-runtime.yaml:28]: That wouldn't work, since we don't install packages through dpkg, pam-auth-update won't be able to iterate packages and find which ones have these templates. This must be emulated instead.

File: slices/libwrap0.yaml
Comment (Meulengracht) [reply slices/libwrap0.yaml:44]: Absolutely, will add a comment for this

File: slices/libpam-runtime.yaml
Comment (cjdcordeiro) [reply slices/libpam-runtime.yaml:28]: Does it make sense to keep this slice then? If the delivered binary is not gonna be functional...

Comment (Meulengracht) [reply slices/libpam-runtime.yaml:28]: That is a good point, I guess it makes sense to remove it completely. I'll leave a comment in its place

File: slices/sysvinit-utils.yaml
Comment (Meulengracht) [reply slices/sysvinit-utils.yaml:23]: done

File: slices/ucf.yaml
Comment (Meulengracht) [reply slices/ucf.yaml]: done

File: slices/libpam-modules-bin.yaml
Comment (Meulengracht) [reply slices/libpam-modules-bin.yaml:1]: Nope, I think some deps have changed between when I created them and now

File: slices/libpam-modules.yaml
Comment (Meulengracht) [reply slices/libpam-modules.yaml]: Removed

File: slices/libpam-runtime.yaml
Comment (Meulengracht) [reply slices/libpam-runtime.yaml]: Removed most core/standard slices and now refer to just whats needed

File: slices/libpam-runtime.yaml
Comment (Meulengracht) [reply slices/libpam-runtime.yaml]: It was, but I have removed the standard slice

File: slices/libpam-modules.yaml
Comment (Meulengracht) [reply slices/libpam-modules.yaml]: Same

File: slices/libselinux1.yaml
Comment (Meulengracht) [reply slices/libselinux1.yaml]: good point, changed

File: slices/libwrap0.yaml
Comment (Meulengracht) [reply slices/libwrap0.yaml]: I think it changed

File: slices/openssh-client.yaml
Comment (Meulengracht) [reply slices/openssh-client.yaml:11]: Passwd is a bit of a special package, because most of the contents are only used by libpam, or user management. I added the package + deps - with comments explaining each slice. I also added a dependency on passwd here

File: slices/openssh-server.yaml
Comment (Meulengracht) [reply slices/openssh-server.yaml:1]: Added!

File: slices/openssh-server.yaml
Comment (Meulengracht) [reply slices/openssh-server.yaml]: Installing just bins now should work, made some slice changes to reflect this now after testing

File: slices/openssh-sftp-server.yaml
Comment (Meulengracht) [reply slices/openssh-sftp-server.yaml]: Refactored the standard slices, should work now

File: slices/sensible-utils.yaml
Comment (Meulengracht) [reply slices/sensible-utils.yaml:16]: Add the missing bin and moved them to their own slice

File: slices/libwrap0.yaml
Comment (Meulengracht) [reply slices/libwrap0.yaml]: As we agreed, this is now a mutation script.

**Comment by Meulengracht** [comment]
@cjdcordeiro I need a re-review

File: slices/libwrap0.yaml
Comment (cjdcordeiro) [inline slices/libwrap0.yaml:53]: nice

File: slices/sensible-utils.yaml
Comment (cjdcordeiro) [inline slices/sensible-utils.yaml:15]:
```suggestion
  extras:
```

this is a name we already use in other pkgs (like python)

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]

File: slices/sensible-utils.yaml
Comment (Meulengracht) [reply slices/sensible-utils.yaml]: fixed!

**Review by cjdcordeiro — APPROVED** [review]

**Pull request merged**
